modules/pam_setquota/pam_setquota.c

```diff
@@ -0,0 +1,273 @@
+  if (retval != PAM_SUCCESS || user == NULL || *(const char *)user == '\0') {
+    pam_syslog(pamh, LOG_NOTICE, "user unknown");
+    return PAM_USER_UNKNOWN;
+  }
+
+  /* Get the password entry */
+  pwd = pam_modutil_getpwnam(pamh, user);
+  if (pwd == NULL) {
+    return PAM_CRED_INSUFFICIENT;
+  }
+
+  /* Check if we should not set quotas for user */
+  if ((pwd->pw_uid < param.start_uid) ||
+      ((param.end_uid >= param.start_uid) && (param.start_uid != 0) &&
+       (pwd->pw_uid > param.end_uid)))
+    return PAM_SUCCESS;
+
+  /* Remove the unnecessary '/' from the end of fs parameter */
+  if (param.fs != NULL) {
+    size_t len = strlen(param.fs);
+    if (len > 1)
+      if (param.fs[len - 1] == '/')
+        param.fs[len - 1] = '\0';
+  }
+
+  /* Find out what device the filesystem is hosted on */
+  if ((fp = setmntent("/etc/mtab", "r")) == NULL) {
+    pam_syslog(pamh, LOG_ERR, "Unable to open /etc/mtab");
+    return PAM_PERM_DENIED;
+  }
+
+  while ((mnt = getmntent(fp)) != NULL) {
+    if (param.fs == NULL) {
+      /* If fs is not specified use filesystem with homedir as default */
+      size_t mnt_len = strlen(mnt->mnt_dir);
+      const char *s;
+      if (mnt_len > match_size &&
+         (s = pam_str_skip_prefix_len(pwd->pw_dir, mnt->mnt_dir, mnt_len)) != NULL &&
+         (s[0] == '\0' || s[0] == '/')) {
+        free(mntdevice);
+        if ((mntdevice = strdup(mnt->mnt_fsname)) == NULL) {
+          pam_syslog(pamh, LOG_CRIT,  "Memory allocation error");
+          endmntent(fp);
+          return PAM_PERM_DENIED;
+        }
+        match_size = mnt_len;
+      }
+    } else if ((strcmp(param.fs, mnt->mnt_dir) == 0) ||
+               (strcmp(param.fs, mnt->mnt_fsname) == 0)) {
+        free(mntdevice);
+        if ((mntdevice = strdup(mnt->mnt_fsname)) == NULL) {
+          pam_syslog(pamh, LOG_CRIT,  "Memory allocation error");
+          endmntent(fp);
+          return PAM_PERM_DENIED;
+        }
+    }
+  }
+
+  endmntent(fp);
+
+  if (mntdevice == NULL) {
+    pam_syslog(pamh, LOG_ERR, "Filesystem not found");
+    return PAM_PERM_DENIED;
+  }
+
+  /* Get limits */
+  if (quotactl(QCMD(Q_GETQUOTA, USRQUOTA), mntdevice, pwd->pw_uid,
+               (void *)&ndqblk) == -1) {
+    pam_syslog(pamh, LOG_ERR, "fail to get limits for user %s : %m",
+               pwd->pw_name);
+    free(mntdevice);
+    return PAM_PERM_DENIED;
+  }
+
+  if (param.debug == 1)
+    debug(pamh, &ndqblk);
+
+  /* Only overwrite if quotas aren't already set or if overwrite is set */
+  if ((ndqblk.dqb_bsoftlimit == 0 && ndqblk.dqb_bhardlimit == 0 &&
+       ndqblk.dqb_isoftlimit == 0 && ndqblk.dqb_ihardlimit == 0) ||
+      param.overwrite == 1) {
+
+    /* Parse new limits */
+    ndqblk.dqb_valid = 0;
+    if (!_pam_parse_dqblk(argc, argv, &ndqblk)) {
+      pam_syslog(pamh, LOG_AUTH | LOG_WARNING, "setting soft and hard limits"
+                                               "separately");
+    }
```

Comment:
I'm not sure what does this warning actually mean.

Reply:
I added a comment lining out the possible problem that could occur.

Reply:
Nothing in the documentation hints at being able to do the following:

```
session required        pam_setquota.so bhardlimit=568000 fs=/home 
session required        pam_setquota.so bsoftlimit=512000 fs=/home 
```
It will work but only if you do it in that exact order, because the softlimit *has* to be lower than the hardlimit. And because it will work (at first) but you are setting yourself up for trouble later, the code warns about it.

Also imagine having the above setup in that way and now you decide to change things:

```
session required        pam_setquota.so bhardlimit=268000 fs=/home overwrite=1
session required        pam_setquota.so bsoftlimit=212000 fs=/home overwrite=1
```
Now it will fail for every user with an existing quota, because now the new hardlimit again is lower than the old softlimit.

Normally no user would do this, because the documentation never really shows this possibility.

Reply:
I have reworded and hopefully improved the warning logged when this configuration is encountered.
Because such a configuration technically works, I don't think it is necessary to exit with an error.